examples/dataflow-csvio/src/main/java/com/google/example/csvio/SortContextualHeadersAndRows.java

```diff
@@ -0,0 +1,427 @@
+    public void process(ProcessContext ctx) {
+      Long headerPosition = ctx.sideInput(this.headerPositionView);
+      Row input = ctx.element();
+      RecordWithMetadataHelper record = RecordWithMetadataHelper.of(input);
+      Long recordNum = record.getRecordNum();
+      if (recordNum < headerPosition) {
+        return;
+      }
+      if (recordNum.equals(headerPosition)) {
+        ctx.output(HEADERS, input);
+      } else {
+        ctx.output(ROWS, input);
+      }
+    }
+  }
+
+  static class FindMatchedHeaderFn extends DoFn<Row, KV<String, Long>> {
+
+    private final PCollectionView<String> headerMatchRegexView;
+
+    FindMatchedHeaderFn(PCollectionView<String> headerMatchRegexView) {
+      this.headerMatchRegexView = headerMatchRegexView;
+    }
+
+    @ProcessElement
+    public void process(ProcessContext ctx) {
+      Row input = ctx.element();
+      RecordWithMetadataHelper record = RecordWithMetadataHelper.of(input);
+      String headerMatchRegex = ctx.sideInput(headerMatchRegexView);
+      if (record.getValue() != null && record.getValue().equals(headerMatchRegex)) {
+        ctx.output(KV.of(record.getResourceId(), record.getRecordNum()));
+      }
+    }
+  }
+
+  /**
+   * The {@link DoFn} responsible for sorting a fixed header position given a with the subsequent
+   * CSV records. The header is mapped to the exact position expected.
+   */
+  static class SortMappedFixedPositionHeadersAndRowsFn extends DoFn<Row, Row> {
+
+    private final PCollectionView<Map<String, Long>> mappedResourceIdHeaderPositionView;
+
+    SortMappedFixedPositionHeadersAndRowsFn(
+        PCollectionView<Map<String, Long>> mappedResourceIdHeaderPositionView) {
+      this.mappedResourceIdHeaderPositionView = mappedResourceIdHeaderPositionView;
+    }
+
+    @ProcessElement
+    public void process(ProcessContext ctx) {
+      Row input = ctx.element();
+      RecordWithMetadataHelper record = RecordWithMetadataHelper.of(input);
+      String resourceId = record.getResourceId();
+      Long recordOffset = record.getRecordNum();
+
+      Map<String, Long> resourceIdHeaderPosition =
+          ctx.sideInput(this.mappedResourceIdHeaderPositionView);
+      if (!resourceIdHeaderPosition.containsKey(resourceId)) {
+        ctx.output(ROWS, input);
+        return;
+      }
+
+      Long headerPosition = resourceIdHeaderPosition.get(resourceId);
+      if (recordOffset < headerPosition) {
+        return;
+      }
+      if (recordOffset.equals(headerPosition)) {
+        ctx.output(HEADERS, input);
+      } else {
+        ctx.output(ROWS, input);
+      }
+    }
+  }
+
+  /**
+   * Determines whether the CSV record {@link ContextualTextIO#read()} output is a non-empty line.
+   */
+  static class IsNotEmptyFn extends SimpleFunction<Row, Boolean> {
+
+    @Override
+    public Boolean apply(Row input) {
+      if (input == null) {
+        return false;
+      }
+      return !RecordWithMetadataHelper.of(input).getValue().isEmpty();
```

Comment:
Can be simplified as 
`return input != null && !RecordWithMetadataHelper.of(input).getValue().isEmpty(); `